File: MatematicaApp/src/main/java/com/sergiocruz/MatematicaPro/activity/MainActivity.java
package com.sergiocruz.MatematicaPro.activity;

import android.content.Intent;
import android.os.Bundle;
import android.os.Handler;
import android.support.design.widget.NavigationView;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentTransaction;
import android.support.v4.view.GravityCompat;
import android.support.v4.widget.DrawerLayout;
import android.support.v7.app.ActionBarDrawerToggle;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.view.MenuItem;
import android.view.View;

import com.sergiocruz.MatematicaPro.R;
import com.sergiocruz.MatematicaPro.fragment.DivisoresFragment;
import com.sergiocruz.MatematicaPro.fragment.FatorizarFragment;
import com.sergiocruz.MatematicaPro.fragment.HomeFragment;
import com.sergiocruz.MatematicaPro.fragment.MDCFragment;
import com.sergiocruz.MatematicaPro.fragment.MMCFragment;
import com.sergiocruz.MatematicaPro.fragment.MultiplosFragment;
import com.sergiocruz.MatematicaPro.fragment.PrimesTableFragment;
import com.sergiocruz.MatematicaPro.fragment.PrimorialFragment;

import static com.sergiocruz.MatematicaPro.helper.MenuHelper.verifyStoragePermissions;

public class MainActivity extends AppCompatActivity {
    // tags used to attach the fragments
    private static final String[] FRAGMENT_TAGS = {"home", "mmc", "mdc", "fatorizar", "divisores", "primes_table", "multiplos", "primorial"};
    private static final Fragment[] FRAGMENTS = {new HomeFragment(), new MMCFragment(), new MDCFragment(), new FatorizarFragment(), new DivisoresFragment(), new PrimesTableFragment(), new MultiplosFragment(), new PrimorialFragment()};
    // index to identify current nav menu item
    public static int navItemIndex = 0;
    Fragment mContent;
    private NavigationView navigationView;
    private DrawerLayout drawer;
    private Toolbar toolbar;
    // toolbar titles respected to selected nav menu item
    private String[] activityTitles;
    // flag to load home fragment when user presses back key
    private boolean shouldLoadHomeFragOnBackPress = true;
    private Handler mHandler;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        setContentView(R.layout.activity_main);
        toolbar = (Toolbar) findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);

        mHandler = new Handler();

        drawer = (DrawerLayout) findViewById(R.id.drawer_layout);
        navigationView = (NavigationView) findViewById(R.id.nav_view);

        // Icones coloridos no menu de gaveta lateral
        navigationView.setItemIconTintList(null);

        // load toolbar titles from string resources
        activityTitles = getResources().getStringArray(R.array.nav_item_activity_titles);

        // initializing navigation menu
        setUpNavigationView();

        if (savedInstanceState == null) {
            loadFragment(0);
        } else if (savedInstanceState != null) {
            //Restore the fragment's instance
            mContent = getSupportFragmentManager().getFragment(savedInstanceState, "mContent");
        }
        verifyStoragePermissions(this);
    }

    @Override
    protected void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);

        //Save the fragment's instance
        //getSupportFragmentManager().putFragment(outState, "mContent", mContent);
    }

    /***
     * Returns respected fragment that user
     * selected from navigation menu
     */
    private void loadFragment(int item_index) {
        //remove dot in menu
        navigationView.getMenu().getItem(navItemIndex).setActionView(null);

        navItemIndex = item_index;

        // selecting appropriate nav menu item
        navigationView.getMenu().getItem(navItemIndex).setChecked(true);
        navigationView.getMenu().getItem(navItemIndex).setActionView(R.layout.menu_dot);

        // set toolbar title
        getSupportActionBar().setTitle(activityTitles[navItemIndex]);

        // if user select the current navigation menu again, don't do anything
        // just close the navigation drawer
        if (getSupportFragmentManager().findFragmentByTag(FRAGMENT_TAGS[navItemIndex]) != null) {
            drawer.closeDrawers();
            return;
        }

        // Sometimes, when fragment has huge data, screen seems hanging
        // when switching between navigation menus
        // So using runnable, the fragment is loaded with cross fade effect
        // This effect can be seen in GMail app
        Runnable mPendingRunnable = new Runnable() {
            @Override
            public void run() {
                // update the main content by replacing fragments
                Fragment fragment = FRAGMENTS[navItemIndex];
                FragmentTransaction fragmentTransaction = getSupportFragmentManager().beginTransaction();
                fragmentTransaction.setCustomAnimations(android.R.anim.fade_in, android.R.anim.fade_out);
                fragmentTransaction.replace(R.id.frame, fragment, FRAGMENT_TAGS[navItemIndex]);
                fragmentTransaction.commitAllowingStateLoss();
            }
        };

        // If mPendingRunnable is not null, then add to the message queue
        if (mPendingRunnable != null) {
            mHandler.post(mPendingRunnable);
        }

        //Closing drawer on item click
        drawer.closeDrawers();

        // refresh toolbar menu
        invalidateOptionsMenu();
    }

    private void setUpNavigationView() {
        //Setting Navigation View Item Selected Listener to handle the item click of the navigation menu
        navigationView.setNavigationItemSelectedListener(new NavigationView.OnNavigationItemSelectedListener() {

            // This method will trigger on item Click of navigation menu
            @Override
            public boolean onNavigationItemSelected(MenuItem menuItem) {

                //remove dot in menu
                navigationView.getMenu().getItem(navItemIndex).setActionView(null);

                //Check to see which item was being clicked and perform appropriate action
                switch (menuItem.getItemId()) {
                    //Replacing the main content with ContentFragment Which is our Inbox View;
                    case R.id.home:
                        navItemIndex = 0;
                        break;
                    case R.id.nav_mmc:
                        navItemIndex = 1;
                        break;
                    case R.id.nav_mdc:
                        navItemIndex = 2;
                        break;
                    case R.id.nav_fatorizar:
                        navItemIndex = 3;
                        break;
                    case R.id.nav_divisores:
                        navItemIndex = 4;
                        break;
                    case R.id.nav_prime_table:
                        navItemIndex = 5;
                        break;
                    case R.id.nav_multiplos:
                        navItemIndex = 6;
                        break;
                    case R.id.nav_primorial:
                        navItemIndex = 7;
                        break;
                    case R.id.nav_settings:
                        startActivity(new Intent(MainActivity.this, SettingsActivity.class));
                        drawer.closeDrawers();
                        return true;
                    case R.id.nav_about:
                        // launch new intent instead of loading fragment
                        startActivity(new Intent(MainActivity.this, AboutActivity.class));
                        drawer.closeDrawers();
                        return true;
                    case R.id.nav_send:
                        // launch new intent instead of loading fragment
                        startActivity(new Intent(MainActivity.this, SendMailActivity.class));
                        drawer.closeDrawers();
                        return true;
                    default:
                        navItemIndex = 0;
                }

                //Checking if the item is in checked state or not, if not make it in checked state
                if (menuItem.isChecked()) {
                    menuItem.setChecked(false);
                } else {
                    menuItem.setChecked(true);
                }
                menuItem.setChecked(true);

                loadFragment(navItemIndex);

                return true;
            }
        });


        ActionBarDrawerToggle actionBarDrawerToggle = new ActionBarDrawerToggle(this, drawer, toolbar, R.string.openDrawer, R.string.closeDrawer) {

            @Override
            public void onDrawerClosed(View drawerView) {
                // Code here will be triggered once the drawer closes as we dont want anything to happen so we leave this blank
                super.onDrawerClosed(drawerView);
            }

            @Override
            public void onDrawerOpened(View drawerView) {
                // Code here will be triggered once the drawer open as we dont want anything to happen so we leave this blank
                super.onDrawerOpened(drawerView);
            }
        };

        //Setting the actionbarToggle to drawer layout
        drawer.addDrawerListener(actionBarDrawerToggle);

        //calling sync state is necessary or else your hamburger icon wont show up
        actionBarDrawerToggle.syncState();
    }

    @Override
    public void onBackPressed() {
        if (drawer.isDrawerOpen(GravityCompat.START)) {
            drawer.closeDrawers();
            return;
        }

        //remove dot from selected item menu
        navigationView.getMenu().getItem(navItemIndex).setActionView(null);

        // This code loads home fragment when back key is pressed
        // when user is in other fragment than home
        if (shouldLoadHomeFragOnBackPress) {
            // checking if user is on other navigation menu
            // rather than home
            if (navItemIndex != 0) {
                loadFragment(0);
                return;
            }
        }

        super.onBackPressed();
    }

    public void mmc(View view) {
        loadFragment(1);
    }

    public void mdc(View view) {
        loadFragment(2);
    }

    public void fatorizar(View view) {
        loadFragment(3);
    }

    public void divisores(View view) {
        loadFragment(4);
    }

    public void primes_table(View view) {
        loadFragment(5);
    }

    public void multiplos(View view) {
        loadFragment(6);
    }

    public void primorial(View view) {
        loadFragment(7);
    }

}

File: MatematicaApp/src/main/java/com/sergiocruz/MatematicaPro/activity/MainActivity.kt
package com.sergiocruz.MatematicaPro.activity

import android.content.Intent
import android.content.pm.PackageManager
import android.os.Bundle
import android.os.Handler
import android.view.View
import androidx.appcompat.app.ActionBarDrawerToggle
import androidx.appcompat.app.AppCompatActivity
import androidx.core.app.ActivityCompat
import androidx.core.view.GravityCompat
import com.sergiocruz.MatematicaPro.R
import com.sergiocruz.MatematicaPro.allPermissionsGranted
import com.sergiocruz.MatematicaPro.fragment.*
import com.sergiocruz.MatematicaPro.getRuntimePermissions
import com.sergiocruz.MatematicaPro.helper.openSettingsFragment
import kotlinx.android.synthetic.main.activity_main.*
import kotlinx.android.synthetic.main.app_bar_main.*

class MainActivity : AppCompatActivity(), ActivityCompat.OnRequestPermissionsResultCallback {
    private lateinit var activityTitles: Array<String>
    private lateinit var mHandler: Handler

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        setContentView(R.layout.activity_main)

        setSupportActionBar(toolbar)
        supportActionBar?.setDisplayShowTitleEnabled(false)

        mHandler = Handler()

        // load toolbar titles from string resources
        activityTitles = resources.getStringArray(R.array.nav_item_activity_titles)

        // initializing navigation menu
        setUpNavigationView()

        if (allPermissionsGranted(this).not()) getRuntimePermissions(this)

        when (savedInstanceState) {
            null -> loadFragment(HomeFragment())
            else -> return
        }

    }

    interface PermissionResultInterface {
        fun onPermissionResult(permitted: Boolean)
    }

    override fun onRequestPermissionsResult(
        requestCode: Int,
        permissions: Array<out String>,
        grantResults: IntArray
    ) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        permissionResultInterface?.onPermissionResult(grantResults[0] == PackageManager.PERMISSION_GRANTED)
    }

    /***
     * Returns respected fragment that user
     * selected from navigation menu
     */
    private fun loadFragment(fragment: BaseFragment) {
        //remove dot in menu
        nav_view.menu.getItem(getCurrentFragmentIndex()).actionView = null

        // selecting appropriate nav menu item
        nav_view.menu.getItem(fragment.pageIndex).isChecked = true
        nav_view.menu.getItem(fragment.pageIndex).setActionView(R.layout.menu_dot)

        // set toolbar title
        toolbarTitle.setText(fragment.title)

        // if user select the current navigation menu again, don't do anything
        // just close the navigation drawer_layout
        if (supportFragmentManager.fragments.getOrNull(0)?.tag == fragment::class.java.simpleName) {
            drawer_layout.closeDrawers()
            return
        }

        // Sometimes, when fragment has huge data, screen seems hanging
        // when switching between navigation menus
        // So using runnable, the fragment is loaded with cross fade effect
        // This effect can be seen in GMail app
        mHandler.post {
            // update the main content by replacing fragments
            val fragmentTransaction = supportFragmentManager.beginTransaction()
            fragmentTransaction.setCustomAnimations(android.R.anim.fade_in, android.R.anim.fade_out)
            fragmentTransaction.replace(R.id.frame, fragment, fragment::class.java.simpleName)
            fragmentTransaction.commitAllowingStateLoss()
        }

        //Closing drawer_layout on item click
        drawer_layout.closeDrawers()

        // refresh toolbar menu
        invalidateOptionsMenu()
    }

    private fun getCurrentFragmentIndex(): Int {
        val fragmentList = supportFragmentManager.fragments
        val stackSize = fragmentList.size
        return (fragmentList.getOrNull(stackSize - 1) as? BaseFragment)?.pageIndex ?: 0
    }


    private fun goToSettingsFragment() {
        nav_view.menu.getItem(getCurrentFragmentIndex()).actionView = null
        nav_view.menu.getItem(SettingsFragment.index).isChecked = true
        nav_view.menu.getItem(SettingsFragment.index).setActionView(R.layout.menu_dot)
        openSettingsFragment()
        drawer_layout.closeDrawers()
    }

    private fun setUpNavigationView() {

        // Icones coloridos no menu de gaveta lateral
        nav_view.itemIconTintList = null

        //Setting Navigation View Item Selected Listener to handle the item click of the navigation menu
        // This method will trigger on item Click of navigation menu
        nav_view.setNavigationItemSelectedListener { menuItem ->

            //remove dot in menu
            nav_view.menu.getItem(getCurrentFragmentIndex()).actionView = null

            when (menuItem.itemId) {
                R.id.home -> loadFragment(HomeFragment())
                R.id.nav_primality -> loadFragment(PrimalityFragment())
                R.id.nav_mmc -> loadFragment(MMCFragment())
                R.id.nav_mdc -> loadFragment(MDCFragment())
                R.id.nav_fatorizar -> loadFragment(FatorizarFragment())
                R.id.nav_divisores -> loadFragment(DivisoresFragment())
                R.id.nav_prime_table -> loadFragment(PrimesTableFragment())
                R.id.nav_multiplos -> loadFragment(MultiplosFragment())
                R.id.nav_primorial -> loadFragment(PrimorialFragment())
                R.id.nav_settings -> goToSettingsFragment()
                R.id.nav_about -> startActivity(Intent(this@MainActivity, AboutActivity::class.java))
                R.id.nav_send -> startActivity(Intent(this@MainActivity, SendMailActivity::class.java))
                else -> loadFragment(HomeFragment())
            }

            //Checking if the item is in checked state or not, if not make it in checked state
            menuItem.isChecked = menuItem.isChecked.not()

            true
        }

        val actionBarDrawerToggle = ActionBarDrawerToggle(this, drawer_layout, toolbar, R.string.openDrawer, R.string.closeDrawer)

        //Setting the actionbarToggle to drawer_layout layout
        drawer_layout.addDrawerListener(actionBarDrawerToggle)

        //calling sync state is necessary or else your hamburger icon wont show up
        actionBarDrawerToggle.syncState()
    }

    override fun onBackPressed() {
        if (drawer_layout.isDrawerOpen(GravityCompat.START)) {
            drawer_layout.closeDrawers()
            return
        }

        //remove dot from selected item menu
        nav_view.menu.getItem(getCurrentFragmentIndex()).actionView = null

        // checking if user is on other navigation menu
        // rather than home
        if (getCurrentFragmentIndex() != 0) {
            loadFragment(HomeFragment())
            return
        }

        super.onBackPressed()
    }

    fun primality(view: View) = loadFragment(PrimalityFragment())

    fun mmc(view: View) = loadFragment(MMCFragment())

    fun mdc(view: View) = loadFragment(MDCFragment())

    fun fatorizar(view: View) = loadFragment(FatorizarFragment())

    fun divisores(view: View) = loadFragment(DivisoresFragment())

    fun primesTable(view: View) = loadFragment(PrimesTableFragment())

    fun multiplos(view: View) = loadFragment(MultiplosFragment())

    fun primorial(view: View) = loadFragment(PrimorialFragment())

    companion object {
        var permissionResultInterface: PermissionResultInterface? = null
    }

}

